### PR TITLE
Fix drop positioning

### PR DESCRIPTION
**Pull request by jcfilben**
#### What does this PR do?
Resolves an issue where drop positioning was not correctly flipping when there was not enough room to display in one direction.

See code comments in DropContainer for a more thorough explanation of what is happening. In cases where the drop has a smaller height than the target and the position is either `top: top` or `bottom: bottom` I decided to not try and flip the drop position. We can change this if we want. I just couldn't think of use cases where we would want this behavior.

This PR covers use cases where the drop doesn't have a scroll within it. Drops that do have a scroll within them don't respond well to the screen size adjusting. The work to address the issues with Drop position when they have a scroll will be handled as a separate issue/PR.

#### How should this be manually tested?
Tested with the existing stories (Drop/All not stretched, List/Action, Form/Controlled) as well as this story:
```
import React from 'react';

import { grommet, Box, Button, Grommet, Tip } from 'grommet';

export const Simple = () => (
  // Uncomment <Grommet> lines when using outside of storybook
  // <Grommet theme={...}>
  <Box align="center" justify="center">
    <Box background="pink" height="large">
      spacer
    </Box>
    <Tip
      defaultVisible
      content={
        <Box width="30px" background="green">
          tip
        </Box>
      }
      dropProps={{
        align: { bottom: 'bottom' },
      }}
      plain
    >
      <Button label="action" size="large" />
    </Tip>
    <Box background="pink" height="large">
      spacer
    </Box>
  </Box>
  //</Grommet>
);

Simple.args = {
  full: true,
};

Simple.parameters = {
  chromatic: { disable: true },
};

export default {
  title: 'Controls/Tip/Simple',
};


```

#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `asFragment()` is used for snapshot testing.

#### What are the relevant issues?
Closes https://github.com/grommet/grommet/issues/6997

#### Do the grommet docs need to be updated?
no
#### Should this PR be mentioned in the release notes?
yes
#### Is this change backwards compatible or is it a breaking change?
backwards compatible

### CODE REVIEW

File: src/js/components/Drop/DropContainer.js [hunk 1]
```diff
@@ -209,23 +209,61 @@ const DropContainer = forwardRef(
             less room in the other direction leave the Drop in its current
             position. */
           if (
+            /* The bottom of the drop is aligned to the top of the target
```
Comment (britt6612) [inline src/js/components/Drop/DropContainer.js:212]: I wonder if we can simplify I know you added to help but maybe just saying if  `align.bottom === 'top'` then the top of the drop is at the edge of the viewport or beyond....

File: src/js/components/Drop/DropContainer.js [hunk 2]
```diff
@@ -209,23 +209,61 @@ const DropContainer = forwardRef(
...
+            above the target change the drop's position so that the bottom 
+            of the drop is aligned to the top of the target. */
+            responsive &&
+            align.top === 'bottom' &&
+            targetRect.bottom + containerRect.height >= windowHeight &&
+            targetRect.top - containerRect.height > 0
           ) {
             bottom = targetRect.top;
             maxHeight = bottom;
+          } else if (
```
Comment (britt6612) [inline src/js/components/Drop/DropContainer.js:252]: code looks good though these are the same checks I was thinking when I was looking at this breaking everything apart thank you 👍

File: src/js/components/Menu/Menu.js [hunk 3]
```diff
@@ -439,7 +439,9 @@ const Menu = forwardRef((props, ref) => {
               background={dropBackground || theme.menu.background}
               {...passThemeFlag}
             >
-              {alignControlMirror === 'top' && align.top === 'top'
+              {alignControlMirror === 'top' &&
+              align.bottom !== 'top' &&
```
Comment (britt6612) [inline src/js/components/Menu/Menu.js:443]: why was this added?

**Comment by taysea** [comment]
In this scenario, it feels like it would be better for the Drop to stay at the top and have a scroll rather than flip below. Should we consider this kind of case?

Example code:

```
import React, { useState } from 'react';

import { Box, Text } from 'grommet';
import { SelectMultiple } from '../SelectMultiple';

const defaultOptions = [
  'Apple',
  'Orange',
  'Banana',
  'Grape',
  'Melon',
  'Strawberry',
  'Kiwi',
  'Mango',
  'Raspberry',
  'Rhubarb',
];

export const Default = () => {
  const [options, setOptions] = useState(defaultOptions);
  const [valueMultiple, setValueMultiple] = useState([]);

  return (
    // Uncomment <Grommet> lines when using outside of storybook
    // <Grommet theme={...}>
    <Box fill align="center" justify="end" pad="large" gap="large">
      <Text>SelectMultiple Default</Text>
      <SelectMultiple
        value={valueMultiple}
        placeholder="Select"
        options={options}
        onSearch={(text) => {
          // The line below escapes regular expression special characters:
          // [ \ ^ $ . | ? * + ( )
          const escapedText = text.replace(/[-\\^$*+?.()|[\]{}]/g, '\\$&');

          // Create the regular expression with modified value which
          // handles escaping special characters. Without escaping special
          // characters, errors will appear in the console
          const exp = new RegExp(escapedText, 'i');
          setOptions(defaultOptions.filter((o) => exp.test(o)));
        }}
        onClose={() => setOptions(defaultOptions)}
        onChange={({ value }) => {
          setValueMultiple(value);
        }}
        // dropHeight="small"
      />
    </Box>
    // </Grommet>
  );
};

Default.parameters = {
  chromatic: { disable: true },
};

Default.args = {
  full: true,
};

export default {
  title: 'Input/SelectMultiple/Default',
};
```

https://github.com/user-attachments/assets/2710ea8f-7d77-46df-8e75-b7b7b8ab92ba

File: src/js/components/Drop/DropContainer.js (hunk 1 above)
Comment (jcfilben) [reply src/js/components/Drop/DropContainer.js:212]: Yeah I can go back and simplify the comments

File: src/js/components/Menu/Menu.js (hunk 3 above)
Comment (jcfilben) [reply src/js/components/Menu/Menu.js:443]: To ensure that when `align.bottom !== 'top'` or `align.top !== 'bottom'` we don't add the controlMirror

**Comment by jcfilben** [comment]
> In this scenario, it feels like it would be better for the Drop to stay at the top and have a scroll rather than flip below. Should we consider this kind of case?

Yeah I think this work might fit better in with the exploration on drops without set heights, and how to manage scroll and positioning in those situations. Let me know what you think or if you think it should be in this PR

File: src/js/components/Drop/DropContainer.js (hunk 1 above)
Comment (taysea) [reply src/js/components/Drop/DropContainer.js:212]: Screenshot 2024-09-24 at 3 14 09 PM

@jcfilben @britt6612 Found myself still trying to parse the conditions as I reviewed. Here's a suggestion of a different way to comment -- basically inline with each condition instead of above. Do you find this readable/helpful? Or do you prefer with the comments above each block?

**Comment by taysea** [comment]
> > In this scenario, it feels like it would be better for the Drop to stay at the top and have a scroll rather than flip below. Should we consider this kind of case?
> 
> Yeah I think this work might fit better in with the exploration on drops without set heights, and how to manage scroll and positioning in those situations. @taysea Let me know what you think or if you think it should be in this PR

Thanks for clarifying, yes I'm comfortable handling in a separate PR and keeping this PR scoped to Drops with defined heights/those that do not use InfiniteScroll internally.

File: src/js/components/Drop/DropContainer.js [hunk 4]
```diff
@@ -209,23 +209,61 @@ const DropContainer = forwardRef(
...
+            align.bottom === 'top' &&
+            targetRect.top - containerRect.height <= 0 &&
+            targetRect.bottom + containerRect.height < windowHeight
           ) {
             top = targetRect.bottom;
             maxHeight = top;
           } else if (
+            /* The top of the drop is aligned to the top of the target, the 
+            bottom of the drop is either at the bottom edge of the viewport
+            or beyond it, and the height of the drop is larger than the target.
```
Comment (taysea) [inline src/js/components/Drop/DropContainer.js:227]: I know we discussed it but can you remind me why it matters that the height of the drop is larger than the target?

File: src/js/components/Drop/DropContainer.js [hunk 5]
```diff
@@ -209,23 +209,61 @@ const DropContainer = forwardRef(
             less room in the other direction leave the Drop in its current
             position. */
           if (
+            /* The bottom of the drop is aligned to the top of the target 
+            and the top of the drop is either at the top edge of the viewport
+            or beyond it. If there is room to display the drop below the target
+            change the drop's position so that the top of the drop is aligned 
+            to the bottom of the target. */
             responsive &&
```
Comment (taysea) [inline src/js/components/Drop/DropContainer.js:217]: The top few conditions check for `responsive &&` then the bottom ones don't. Would it be better to write this logic as:
```
if (responsive) {
   if (
      align.bottom === 'top' && 
      // drop is overflowing above window
      targetRect.top - containerRect.height <= 0 && 
      // there's space below the target 
      targetRect.bottom + containerRect.height < windowHeight
   ) { 
       ... 
   } else if () {
   } ...
} else if (align.top === 'top') {...}
...
```
just would save a few condition checks for the non-responsive cases

**Review by taysea — COMMENTED** [review]
Dropped some cleanup comments but the behavior is looking good.

File: src/js/components/Drop/DropContainer.js (hunk 1 above)
Comment (jcfilben) [reply src/js/components/Drop/DropContainer.js:212]: ^ I like this format, switched to this in latest commit

File: src/js/components/Drop/DropContainer.js (hunk 4 above)
Comment (jcfilben) [reply src/js/components/Drop/DropContainer.js:227]: If the height of the drop is smaller than the target and the drop is positioned so that it is displayed over the target then it doesn't do much good to try and reposition the drop unless we had a really large target. This check just prevents unnecessary re-positioning that doesn't really help make the drop more visible

File: src/js/components/Drop/DropContainer.js (hunk 5 above)
Comment (jcfilben) [reply src/js/components/Drop/DropContainer.js:217]: If we restructure this then we run into issues where if responsive is true but it doesn't meet any of those top 4 conditions then it will never reach the other else if statements or the else statement at the end.

Comment (taysea) [reply src/js/components/Drop/DropContainer.js:212]: Ah thank you for clarifying.

**Review by taysea — APPROVED** [review]
LGTM, and super helpful additions with the comments. Thank you!

**Review by britt6612 — APPROVED** [review]
Thanks for cleaning up comments 🙌
